libcontainer/nsenter/nsexec.c

```diff
@@ -77,12 +78,18 @@ static int child_func(void *_arg)
     longjmp(*arg->env, 1);
 }
 
-static int clone_parent(jmp_buf *env, int flags) __attribute__((noinline));
-static int clone_parent(jmp_buf *env, int flags)
+static int clone_parent(jmp_buf *env, int flags, bool delay_ipc_unshare) __attribute__((noinline));
+static int clone_parent(jmp_buf * env, int flags, bool delay_ipc_unshare)
 {
 	struct clone_arg ca;
 	int		 child;
 
+	// Don't clone into NEWIPC at the same time as cloning into NEWUSER.
+	// This way we can ensure that NEWIPC namespace belongs to the root in new user namespace.
+	if (delay_ipc_unshare) {
```

Comment:
This behaviour is guaranteed by most recent Linux kernels (when you set `CLONE_NEW<namespace>` as well as `CLONE_NEWUSER`, the user namespace is created first). However, if this is a problem on older RedHat kernels then the proper fix should use `unshare` for the user namespace and then use clone for the rest of the namespaces. This code **already exists** in my rootless container PR, but I'd be happy to port the code to #950 (where a bunch of other `nsenter` cleanups are happening).